models/runner/functions.py

```diff
@@ -118,6 +130,35 @@ def parse_exif(df):
 
     return result
 
+def add_station_coords(df, grid_file):
+    """Add station coordinates matched from the grid file"""
+    # Grid file column name --> output column name mapping.
+    columns = {
+        "locationID": "station",
+        "Longitude": "grid_file_long",
+        "Latitude": "grid_file_lat",
+    }
+    if grid_file is None:
+        # Use empty data frame - as if the grid file had no stations.
+        stations = pd.DataFrame(columns=columns.values())
+    else:
+        stations = pd.read_csv(grid_file)
+        stations.rename(columns=columns, inplace=True, errors="raise")
+    return pd.merge(df, stations, how="left", on="station")
+
+def add_output_coords(df):
+    """Add output coordinates determined on best-effort basis"""
+    def f(row):
+        # We assume that longitude is present iff latiitude is present.
+        if pd.notnull(row["exif_gps_long"]):  # Prefer Exif.
```

Comment:
Have you considered requiring both to be filled?

Reply:
That's a good suggestion! The source columns (`exif_gps_*` and `grid_file_*`) are present in the output CSV anyway, so it will be easy to investigate what went wrong if the output columns (`coordinates_*`) are missing. Hopefully longitude is indeed present if and only if latitude is present, but input data does have severe inconsistencies sometimes and assuming missing data is probably the best choice in that case.

I will merge this PR after applying a small patch for this piece of code.